### PR TITLE
TSK-1800: add secondary object references

### PR DESCRIPTION
**Pull request by ryzheboka**
https://sonarcloud.io/summary/new_code?branch=TSK-1800&id=ryzheboka_taskana

---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [ ] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

File: rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java [hunk 1]
```diff
@@ -579,6 +624,90 @@ void should_NotGetEmptyAttachmentList_When_GettingTaskWithAttachment() {
...
+  }
+
+  @Test
+  void should_CreateAndDeleteTaskWithObjectReferences_WhenSpecifyingObjectReferences() {
+    TaskRepresentationModel taskRepresentationModel = getTaskResourceSample();
+    ObjectReferenceRepresentationModel obj0 = getSampleSecondaryObjectReference("0");
+    obj0.setTaskId(taskRepresentationModel.getTaskId());
+    ObjectReferenceRepresentationModel obj1 = getSampleSecondaryObjectReference("1");
+    obj1.setTaskId(taskRepresentationModel.getTaskId());
+    // System.out.println(obj1.getTaskId() + " " + obj0.getTaskId());
```
Comment (gitgoodjhe) [inline rest/taskana-rest-spring/src/test/java/pro/taskana/task/rest/TaskControllerIntTest.java:688]: please delete this

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]

File: common/taskana-common-data/src/main/resources/sql/sample-data/object-reference.sql [hunk 2]
```diff
@@ -1,3 +1,4 @@
-INSERT INTO OBJECT_REFERENCE VALUES ('1', 'Company1', 'System1', 'Instance1', 'Type1', 'Value1');
-INSERT INTO OBJECT_REFERENCE VALUES ('2', 'Company2', 'System2', 'Instance2', 'Type2', 'Value2');
-INSERT INTO OBJECT_REFERENCE VALUES ('3', 'Company3', 'System3', 'Instance3', 'Type3', 'Value3');
+INSERT INTO OBJECT_REFERENCE VALUES ('ID1', 'TKI:000000000000000000000000000000000000', 'Company1', 'System1', 'Instance1', 'Type1', 'Value1');
```
Comment (gitgoodjhe) [inline common/taskana-common-data/src/main/resources/sql/sample-data/object-reference.sql:1]: Oh no do we have the formatter issue again? :/

File: .run/Start Frontend.run.xml [hunk 3]
```diff
@@ -1,12 +1,12 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Start Frontend" type="js.build_tools.npm">
-    <package-json value="$PROJECT_DIR$/web/package.json" />
-    <command value="run" />
+    <package-json value="$PROJECT_DIR$/web/package.json"/>
+    <command value="run"/>
```
Comment (gitgoodjhe) [inline .run/Start Frontend.run.xml:4]: formatting issues?

File: common/taskana-common-data/src/main/resources/sql/test-data/object-reference.sql [hunk 4]
```diff
@@ -1,3 +1,3 @@
-INSERT INTO OBJECT_REFERENCE VALUES ('1', 'Company1', 'System1', 'Instance1', 'Type1', 'Value1');
-INSERT INTO OBJECT_REFERENCE VALUES ('2', 'Company2', 'System2', 'Instance2', 'Type2', 'Value2');
-INSERT INTO OBJECT_REFERENCE VALUES ('3', 'Company3', 'System3', 'Instance3', 'Type3', 'Value3');
+INSERT INTO OBJECT_REFERENCE VALUES ('1', 'TaskId1', 'Company1', 'System1', 'Instance1', 'Type1', 'Value1');
```
Comment (gitgoodjhe) [inline common/taskana-common-data/src/main/resources/sql/test-data/object-reference.sql:1]: Oh no do we have the formatter issue again? :/

File: lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java [hunk 5]
```diff
@@ -38,133 +39,176 @@
...
+   * @return the planned {@link Instant}
    */
   Instant getPlanned();
 
   /**
-   * Gets the time when when the surrounding process was started.
+   * Gets the time when the surrounding process was started.
    *
-   * @return the received Instant
+   * @return the received {@link Instant}
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java:77]: I don't think we link Instants, since it's not an entity of us, but you can leave this for now :)

File: lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java [hunk 6]
```diff
@@ -38,133 +39,176 @@
...
-   * Gets the domain of the task.
+   * Gets the secondary {@link ObjectReference}s of the Task.
    *
-   * @return the task's domain
+   * @return the Task's secondary {@link ObjectReference}s
+   */
+  List<ObjectReference> getSecondaryObjectReferences();
+
+  /**
+   * Add an {@link ObjectReference} the list of secondary {@link ObjectReference}s.<br>
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java:152]:
```suggestion
   * Add an {@link ObjectReference} to the list of secondary {@link ObjectReference}s.<br>
```

File: lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java [hunk 7]
```diff
@@ -38,133 +39,176 @@
...
+   * NULL will be ignored and an object reference with the same ID will be replaced by the newer
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/models/TaskSummary.java:153]:
```suggestion
   * NULL will be ignored and an ObjectReference with the same ID will be replaced by the newer
```

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/ObjectReferenceHandler.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    List<String> newObjectReferencesIds = new ArrayList<>();
+    if (newObjectReferences != null && !newObjectReferences.isEmpty()) {
+      newObjectReferencesIds =
+          newObjectReferences.stream().map(ObjectReference::getId).collect(Collectors.toList());
+    }
+    List<ObjectReference> oldObjectReferences = oldTaskImpl.getSecondaryObjectReferences();
+    if (oldObjectReferences != null && !oldObjectReferences.isEmpty()) {
+      final List<String> newAttIds = newObjectReferencesIds;
+      oldObjectReferences.forEach(
+          a -> {
```
[the remark this answers is not in the document]
Comment (gitgoodjhe) [reply lib/taskana-core/src/main/java/pro/taskana/task/internal/ObjectReferenceHandler.java:124]: you don't want to change the "a" to "o"? :) a was probbaly meant for attachment, while o could be objectreference...I also wouldn't mind if you named it objectReference

File: lib/taskana-core/src/test/java/acceptance/task/DeleteTaskWithSorAccTest.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+import pro.taskana.task.internal.ObjectReferenceMapper;
+import pro.taskana.task.internal.builder.ObjectReferenceBuilder;
+import pro.taskana.task.internal.builder.TaskBuilder;
+import pro.taskana.workbasket.api.WorkbasketPermission;
+import pro.taskana.workbasket.api.WorkbasketService;
+import pro.taskana.workbasket.api.models.WorkbasketSummary;
+import pro.taskana.workbasket.internal.builder.WorkbasketAccessItemBuilder;
+
+/**
+ * Acceptance test for all "delete task" scenarios that involve secondary {@link ObjectReference}.
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/test/java/acceptance/task/DeleteTaskWithSorAccTest.java:32]:
```suggestion
 * Acceptance test for all "delete task" scenarios that involve secondary {@link ObjectReference}s.
```

File: lib/taskana-core/src/test/java/acceptance/task/TaskQueryImplAccTest.java [hunk 10]
```diff
@@ -2429,6 +2431,534 @@ void should_ApplyFilter_When_QueryingForNotWithin() {
...
+                .workbasketIdIn(wb.getId())
+                .sorSystemInstanceIn("FirstSystemInstance")
+                .list();
+
+        assertThat(list).containsExactlyInAnyOrder(taskSummary1, taskSummary3);
+      }
+
+      @WithAccessId(user = "user-1-1")
+      @Test
+      void should_ApplyFilter_When_QueryingForTypeLike() {
```
[the remark this answers is not in the document]
Comment (gitgoodjhe) [reply lib/taskana-core/src/test/java/acceptance/task/TaskQueryImplAccTest.java:2936]: These test names are still wrong

File: lib/taskana-core/src/test/java/acceptance/task/UpdateTaskWithSorAccTest.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+import pro.taskana.task.api.models.ObjectReference;
+import pro.taskana.task.api.models.Task;
+import pro.taskana.task.internal.builder.ObjectReferenceBuilder;
+import pro.taskana.task.internal.builder.TaskBuilder;
+import pro.taskana.workbasket.api.WorkbasketPermission;
+import pro.taskana.workbasket.api.WorkbasketService;
+import pro.taskana.workbasket.api.models.WorkbasketSummary;
+import pro.taskana.workbasket.internal.builder.WorkbasketAccessItemBuilder;
+
+/** Acceptance test for "update task" scenarios that involve secondary {@link ObjectReference}. */
```
Comment (gitgoodjhe) [inline lib/taskana-core/src/test/java/acceptance/task/UpdateTaskWithSorAccTest.java:26]:
```suggestion
/** Acceptance test for "update task" scenarios that involve secondary {@link ObjectReference}s. */
```

**Review by gitgoodjhe — CHANGES_REQUESTED** [review]
